src/ILCompiler.Compiler/src/Compiler/DependencyAnalysis/InstanceFieldLayoutNode.cs

```diff
@@ -0,0 +1,212 @@
+
+using System;
+
+using Internal.NativeFormat;
+using Internal.Text;
+using Internal.TypeSystem;
+
+using Debug = System.Diagnostics.Debug;
+using TypeFlags = Internal.TypeSystem.TypeFlags;
+
+namespace ILCompiler.DependencyAnalysis
+{
+    /// <summary>
+    /// Represents a hashtable that describes instance field layout of select types.
+    /// Instance field layout information includes information about offsets and types of instance fields.
+    /// </summary>
+    internal sealed class InstanceFieldLayoutNode : ObjectNode, ISymbolDefinitionNode
+    {
+        private readonly ObjectAndOffsetSymbolNode _endSymbol;
+        private readonly ExternalReferencesTableNode _externalReferences;
+
+        public InstanceFieldLayoutNode(ExternalReferencesTableNode externalReferences)
+        {
+            _endSymbol = new ObjectAndOffsetSymbolNode(this, 0, "__instanceFieldLayout_End", true);
+            _externalReferences = externalReferences;
+        }
+
+        public ISymbolNode EndSymbol => _endSymbol;
+
+        public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
+        {
+            sb.Append(nameMangler.CompilationUnitPrefix).Append("__instanceFieldLayout");
+        }
+        public int Offset => 0;
+        public override bool IsShareable => false;
+
+        public override ObjectNodeSection Section => _externalReferences.Section;
+
+        public override bool StaticDependenciesAreComputed => true;
+
+        protected override string GetName(NodeFactory factory) => this.GetMangledName(factory.NameMangler);
+
+        public static void AddDependenciesDueToEETypePresence(ref DependencyList dependencyList, NodeFactory factory, TypeDesc type)
+        {
+            if (factory.Target.Abi == TargetAbi.ProjectN)
+                return;
+
+            // The hash table will reference the types of all instance field if this type qualifies
+            if (NeedsFieldLayoutInformation(type))
+            {
+                dependencyList = dependencyList ?? new DependencyList();
+
+                foreach (FieldDesc field in type.GetFields())
+                {
+                    if (field.IsStatic)
+                        continue;
+
+                    TypeDesc fieldType = NormalizeFieldType(field.FieldType);                    
+                    dependencyList.Add(factory.ConstructedTypeSymbol(fieldType), "Instance field layout");
+                }
+            }
+        }
+
+        public static bool NeedsFieldLayoutInformation(TypeDesc type)
+        {
+            // We need field layout information for boxable valuetypes that cannot be compared
+            // bit-by-bit.
+            //
+            // One might think that the presence of Equals/GetHashCode overrides on the type
+            // would mean we no longer need this information, but we have seen customer code
+            // that does:
+            //
+            // public override bool Equals(object obj) => base.Equals(obj);
```

Comment:
Funny. This make me wonder: can't the compiler simply replace the `base.Equals(obj)` call with code that does the right thing?

Reply:
What would it replace it with though? It would have to be something that also works for annoying things like:

```csharp
class Program
{
    struct Foo
    {
        public Func<int> CreateDelegate() => base.GetHashCode;
        public override int GetHashCode() => 0;
    }

    static void Main()
    {
        Console.WriteLine(default(Foo).GetHashCode());
        Console.WriteLine(default(Foo).CreateDelegate()());
    }
}
```

Reply:
> What would it replace it with though?

With code that does the same thing as `base.Equals(object o)` but without needing to know the field layout at runtime.

In your new example you show a `struct` that doesn't even override `Equals`. I don't know if it's possible but it would be interesting if ILC could automatically override `Equals`.

Reply:
ProjectN does roughly that. However, the generated method bodies can add up to a lot of space in return for the increased efficiency. That compiler actually includes a bit of analysis of which types are likely to have Equals/GetHashCode called and generates bodies for them and then falls back to the base class for cases that it doesn't think are likely (and for cases where it can prove there's no call, it doesn't generate code or metadata). Hopefully we could use something like that here at some point.

Reply:
ProjectN also needs to rewrite direct calls to `System.ValueType::Equals/GetHashCode` to call the injected method, but that code is sketchy and we would need to port it into RyuJIT (and all the other codegen backends) if we ever wanted to take the path of synthesizing the overloads in the compiler (like we do in Project N). I'm not a huge fan.

I haven't measured it yet, but I think the code I'm adding is going to be faster than CoreCLR (except for the part where CoreCLR caches the "can use memcmp?" part and we don't). I'm generally fine with everything that is at least as fast as CoreCLR.